concent_api/api-test.py

```diff
@@ -1,72 +1,152 @@
+CONCENT_PRIVATE_KEY = b'l\xcdh\x19\xeb$>\xbcG\xa1\xc7v\xe8\xd7o\x0c\xbf\x0e\x0fM\x89lw\x1e\xd7K\xd6Hnv$\xa2'
 
 DEFAULT_HEADERS = {
-    'Content-Type':              'application/json',
-    'concent-client-public-key': '85cZzVjahnRpUBwm0zlNnqTdYom1LF1P1WNShLg17cmhN2UssnPrCjHKTi5susO3wrr/q07eswumbL82b4HgOw==',
+    'Content-Type':              'application/octet-stream',
+    'concent-client-public-key': b64encode(CONCENT_PUBLIC_KEY).decode('ascii'),
 }
 
 
 def create_data(message_type, task_id):
+    assert task_id > 0
+
     current_time = int(datetime.datetime.now().timestamp())
-    data = {
-        "type":         message_type,
-        "timestamp":    current_time,
-        "message_task_to_compute": {
-            "type":         "MessageTaskToCompute",
-            "timestamp":    current_time,
-            "task_id":      task_id,
-            "deadline":     current_time + 6000
-        }
-    }
-    return data
-
-
-def create_reject_data(message_type, task_id):
+    message_task_to_compute = MessageTaskToCompute(
+        timestamp = current_time,
+        task_id = task_id,
+        deadline = current_time + 6000,
+    )
+
+    message_task_to_compute_dumped = dump(
+        message_task_to_compute,
+        CONCENT_PRIVATE_KEY,
+        CONCENT_PUBLIC_KEY,
+    )
+
+    if message_type == 'MessageForceReportComputedTask':
+        return MessageForceReportComputedTask(
+            timestamp = current_time,
+            message_task_to_compute = message_task_to_compute_dumped
+        )
+
+    elif message_type == 'MessageAckReportComputedTask':
+        return MessageAckReportComputedTask(
+            timestamp = current_time,
+            message_task_to_compute = message_task_to_compute_dumped,
+        )
+    else:
+        assert False, 'Invalid message type'
+
+    return None
+
+
+def create_reject_data(task_id):
+    assert task_id > 0
+
     current_time = int(datetime.datetime.now().timestamp())
-    data = {
-        "type":         "MessageRejectReportComputedTask",
-        "timestamp":    current_time,
-        "reason":       "cannot-compute-task",
-        "message_cannot_commpute_task": {
-            "type":         "MessageCannotComputeTask",
-            "timestamp":    current_time,
-            "reason":       "provider-quit",
-            "task_id":      task_id
-        }
-    }
-    return data
+    message_cannot_compute_task = MessageCannotComputeTask(
+        timestamp = current_time,
+        reason = 'provider-quit',
+        task_id = task_id,
+    )
+
+    message_cannot_compute_task_dumped = dump(
+        message_cannot_compute_task,
+        CONCENT_PRIVATE_KEY,
+        CONCENT_PUBLIC_KEY,
+
+    )
+
+    return MessageRejectReportComputedTask(
+        timestamp = current_time,
+        reason = 'cannot-compute-task',
+        message_cannot_compute_task = message_cannot_compute_task_dumped,
+    )
+
+
+def print_golem_message(message, indent = 4):
+    assert isinstance(message, Message)
```

Comment:
I see you are a great fan of asserts :) Asserts are ok in general, but I think we want more controlled flow here, with rising and catching exceptions. @cameel what do you think ?

Reply:
See my comments above.

Yes, we use `assert` a lot because it's a great tool to ensure that the assumptions you make are valid. We never use them to control the program flow though. If there's any chance that an assertion will be false even though the code is pefectly valid (e.g. due to user's input, network conditions, hardware failure, etc.) it should indeed be replaced with a standard `if` and raising an exception.

This is not the case here. The function is deliberately written to handle only messages that derive from the `Message` type. Any other case is not even worth coding for and it's the responsibility of the caller to ensure that this precondition holds. E.g. if one wanted to use this function with data where the type is not known, it's his job to check the type and call the function only when he knows it's a `Message`.

This is perfectly appropriate in internal code where you have total control over what arguments the function is called with. It lets you not worry about cases that will never happen but also catch them at the implementation stage if they do happen due to your own mistake.